src/librbd/io/CopyupRequest.cc

```diff
@@ -101,399 +122,480 @@ CopyupRequest<I>::~CopyupRequest() {
+    CopyupRequest<I>,
+    &CopyupRequest<I>::handle_read_from_parent>(
+      this, util::get_image_ctx(m_image_ctx), AIO_TYPE_READ);
+
+  ldout(cct, 20) << "oid=" << m_oid << ", "
+                 << "completion=" << comp << ", "
+                 << "extents=" << m_image_extents
+                 << dendl;
+  ImageRequest<I>::aio_read(m_image_ctx->parent, comp,
+                            std::move(m_image_extents),
+                            ReadResult{&m_copyup_data}, 0, m_trace);
 }
 
 template <typename I>
-bool CopyupRequest<I>::send_copyup() {
-  ldout(m_ictx->cct, 20) << "oid " << m_oid << dendl;
-  m_state = STATE_COPYUP;
+void CopyupRequest<I>::handle_read_from_parent(int r) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "oid=" << m_oid << ", r=" << r << dendl;
 
-  m_ictx->snap_lock.get_read();
-  ::SnapContext snapc = m_ictx->snapc;
-  m_ictx->snap_lock.put_read();
+  m_lock.Lock();
+  m_copyup_required = is_copyup_required();
+  disable_append_requests();
 
-  std::vector<librados::snap_t> snaps;
+  if (r < 0 && r != -ENOENT) {
+    m_lock.Unlock();
+
+    lderr(cct) << "error reading from parent: " << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  if (!m_copyup_required) {
+    m_lock.Unlock();
+
+    ldout(cct, 20) << "no-op, skipping" << dendl;
+    finish(0);
+    return;
+  }
+  m_lock.Unlock();
+
+  update_object_maps();
+}
+
+template <typename I>
+void CopyupRequest<I>::deep_copy() {
+  auto cct = m_image_ctx->cct;
+  ceph_assert(m_image_ctx->snap_lock.is_locked());
+  ceph_assert(m_image_ctx->parent_lock.is_locked());
+  ceph_assert(m_image_ctx->parent != nullptr);
+
+  m_lock.Lock();
+  m_flatten = is_copyup_required() ? true : m_image_ctx->migration_info.flatten;
+  m_lock.Unlock();
+
+  ldout(cct, 20) << "oid=" << m_oid << ", flatten=" << m_flatten << dendl;
+
+  m_deep_copy = true;
+  auto ctx = util::create_context_callback<
+    CopyupRequest<I>, &CopyupRequest<I>::handle_deep_copy>(this);
+  auto req = deep_copy::ObjectCopyRequest<I>::create(
+    m_image_ctx->parent, m_image_ctx, m_image_ctx->migration_info.snap_map,
+    m_object_no, m_flatten, ctx);
+
+  req->send();
+}
+
+template <typename I>
+void CopyupRequest<I>::handle_deep_copy(int r) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "oid=" << m_oid << ", r=" << r << dendl;
+
+  m_image_ctx->snap_lock.get_read();
+  m_lock.Lock();
+  m_copyup_required = is_copyup_required();
+  if (r == -ENOENT && !m_flatten && m_copyup_required) {
+    m_lock.Unlock();
+    m_image_ctx->snap_lock.put_read();
+
+    ldout(cct, 10) << "restart deep-copy with flatten" << dendl;
+    send();
+    return;
+  }
+
+  disable_append_requests();
+
+  if (r < 0 && r != -ENOENT) {
+    m_lock.Unlock();
+    m_image_ctx->snap_lock.put_read();
+
+    lderr(cct) << "error encountered during deep-copy: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  if (!m_copyup_required && !is_update_object_map_required(r)) {
```

Comment:
I got a crash here [1] though was not able to reproduce after rerunning the job.
```
2019-04-07T13:17:53.711 INFO:tasks.workunit.client.0.smithi185.stdout:[ RUN      ] TestMigration.CloneV1Parent
2019-04-07T13:17:53.837 INFO:tasks.workunit.client.0.smithi185.stdout:resize: 83886080
2019-04-07T13:17:53.937 INFO:tasks.workunit.client.0.smithi185.stdout:write: A 0~10
2019-04-07T13:17:53.969 INFO:tasks.workunit.client.0.smithi185.stdout:snap_create: snap1
2019-04-07T13:17:55.302 INFO:tasks.workunit.client.0.smithi185.stdout:snap_protect: snap1
2019-04-07T13:17:55.516 INFO:tasks.workunit.client.0.smithi185.stdout:migration_prepare
2019-04-07T13:17:56.846 INFO:tasks.workunit.client.0.smithi185.stdout:migration_execute
2019-04-07T13:17:57.077 INFO:tasks.workunit.client.0.smithi185.stderr:/build/ceph-15.0.0-126-g70785ee/src/librbd/io/CopyupRequest.cc: In function 'bool librbd::io::CopyupRequest<ImageCtxT>::is_update_object_map_required(int) [with ImageCtxT = librbd::ImageCtx]' thread 7f10b4ff9700 time 2019-04-07 13:17:57.07608420``
2019-04-07T13:17:57.077 INFO:tasks.workunit.client.0.smithi185.stderr:/build/ceph-15.0.0-126-g70785ee/src/librbd/io/CopyupRequest.cc: 556: FAILED ceph_assert(it != m_image_ctx->migration_info.snap_map.end())
2019-04-07T13:17:57.083 INFO:tasks.workunit.client.0.smithi185.stderr: ceph version 15.0.0-126-g70785ee (70785ee3c80326aaea9db18878e0369e5093571e) octopus (dev)
2019-04-07T13:17:57.083 INFO:tasks.workunit.client.0.smithi185.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x152) [0x7f1114cca0d2]
2019-04-07T13:17:57.083 INFO:tasks.workunit.client.0.smithi185.stderr: 2: (ceph::__ceph_assertf_fail(char const*, char const*, int, char const*, char const*, ...)+0) [0x7f1114cca2ad]
2019-04-07T13:17:57.084 INFO:tasks.workunit.client.0.smithi185.stderr: 3: (librbd::io::CopyupRequest<librbd::ImageCtx>::is_update_object_map_required(int)+0xb2) [0xaa20f2]
2019-04-07T13:17:57.084 INFO:tasks.workunit.client.0.smithi185.stderr: 4: (librbd::io::CopyupRequest<librbd::ImageCtx>::handle_deep_copy(int)+0xd8) [0xaa7988]
2019-04-07T13:17:57.084 INFO:tasks.workunit.client.0.smithi185.stderr: 5: (Context::complete(int)+0x9) [0x752899]
2019-04-07T13:17:57.084 INFO:tasks.workunit.client.0.smithi185.stderr: 6: (librbd::deep_copy::ObjectCopyRequest<librbd::ImageCtx>::finish(int)+0x2a5) [0xa6f225]
2019-04-07T13:17:57.084 INFO:tasks.workunit.client.0.smithi185.stderr: 7: (librbd::deep_copy::ObjectCopyRequest<librbd::ImageCtx>::handle_update_object_map(int)+0x4c) [0xa7155c]
2019-04-07T13:17:57.085 INFO:tasks.workunit.client.0.smithi185.stderr: 8: (boost::detail::function::void_function_obj_invoker1<librbd::deep_copy::ObjectCopyRequest<librbd::ImageCtx>::send_update_object_map()::{lambda(int)#4}, void, int>::invoke(boost::detail::function::function_buffer&, int)+0xc) [0xa7184c]
2019-04-07T13:17:57.085 INFO:tasks.workunit.client.0.smithi185.stderr: 9: (FunctionContext::finish(int)+0x2c) [0x7cf0bc]
2019-04-07T13:17:57.085 INFO:tasks.workunit.client.0.smithi185.stderr: 10: (Context::complete(int)+0x9) [0x752899]
2019-04-07T13:17:57.085 INFO:tasks.workunit.client.0.smithi185.stderr: 11: (librbd::ObjectMap<librbd::ImageCtx>::handle_detained_aio_update(librbd::BlockGuardCell*, int, Context*)+0x1ed) [0x892a5d]
2019-04-07T13:17:57.085 INFO:tasks.workunit.client.0.smithi185.stderr: 12: (FunctionContext::finish(int)+0x2c) [0x7cf0bc]
2019-04-07T13:17:57.085 INFO:tasks.workunit.client.0.smithi185.stderr: 13: (librbd::AsyncRequest<librbd::ImageCtx>::finish_and_destroy(int)+0x3c) [0xa4fe4c]
2019-04-07T13:17:57.086 INFO:tasks.workunit.client.0.smithi185.stderr: 14: (librbd::object_map::UpdateRequest<librbd::ImageCtx>::handle_update_object_map(int)+0x332) [0x9eef72]
2019-04-07T13:17:57.086 INFO:tasks.workunit.client.0.smithi185.stderr: 15: (()+0x8b09d) [0x7f111dfe109d]
2019-04-07T13:17:57.086 INFO:tasks.workunit.client.0.smithi185.stderr: 16: (()+0x4f269) [0x7f111dfa5269]
2019-04-07T13:17:57.086 INFO:tasks.workunit.client.0.smithi185.stderr: 17: (Finisher::finisher_thread_entry()+0x16e) [0x7f1114d11e1e]
2019-04-07T13:17:57.086 INFO:tasks.workunit.client.0.smithi185.stderr: 18: (()+0x76ba) [0x7f111dd406ba]
2019-04-07T13:17:57.086 INFO:tasks.workunit.client.0.smithi185.stderr: 19: (clone()+0x6d) [0x7f111472b41d]
2019-04-07T13:18:00.772 INFO:tasks.workunit.client.0.smithi185.stderr:Aborted (core dumped)
```
This might be not related, probably the refactoring just has made it easier to trigger? And right now I can't understand how it  could happen. The assertion failure suggests that at the moment when `is_update_object_map_required` was called the `migration_info` was empty (we do not check this in `handle_deep_copy`), on the other hand I can't see how it could become empty here for this particular test. I suppose this copyup was called by the migration request, and `migration_info` should be cleared only after it is complete?


Anyway, probably we need to add the check that the image is still in migrating state in `handle_deep_copy`, because the `CopyupRequest` could be triggered by a concurrent write request  and the migration could finish earlier?

[1] http://qa-proxy.ceph.com/teuthology/trociny-2019-04-07_08:08:37-rbd-wip-mgolub-testing-distro-basic-smithi/3820282/teuthology.log

Reply:
I appended a commit to address this possible race